chainerchem/dataset/preprocessors/weavenet_preprocessor.py

```diff
@@ -1,17 +1,166 @@
+import os
+
 import numpy
 from rdkit import Chem
+from rdkit import RDConfig
+from rdkit.Chem import AllChem
+from rdkit.Chem import ChemicalFeatures
 
 from chainerchem.dataset.preprocessors.common \
     import construct_atomic_number_array
 from chainerchem.dataset.preprocessors.common import type_check_num_atoms
 from chainerchem.dataset.preprocessors.mol_preprocessor import MolPreprocessor
 
 
-MAX_DISTANCE = 7
-# TODO(Nakago): Review default size
-DEFAULT_NUM_MAX_ATOMS = 20  # 40
+ATOM = ['H', 'C', 'N', 'O', 'S', 'Cl', 'Br', 'F', 'P', 'I']
+MAX_DISTANCE = 2  # 7
+DEFAULT_NUM_MAX_ATOMS = 20  # 60  # paper
+
+
+# --- Atom feature extraction ---
+def construct_atom_type_vector(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    n_atom_type = len(ATOM)
+    n_atom = mol.GetNumAtoms()
+    atom_type_vector = numpy.zeros((num_max_atoms, n_atom_type),
+                                   dtype=numpy.float32)
+    for i in range(n_atom):
+        a = mol.GetAtomWithIdx(i)
+        atom_idx = ATOM.index(a.GetSymbol())
+        atom_type_vector[i, atom_idx] = 1.0
+    return atom_type_vector
+
+
+def construct_formal_charge_vec(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    n_atom = mol.GetNumAtoms()
+    formal_charge_vec = numpy.zeros((num_max_atoms, 1), dtype=numpy.float32)
+    for i in range(n_atom):
+        a = mol.GetAtomWithIdx(i)
+        formal_charge_vec[i, 0] = a.GetFormalCharge()
+    return formal_charge_vec
+
+
+def construct_hybridization_vec(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    n_atom = mol.GetNumAtoms()
+    hybridization_vec = numpy.zeros((num_max_atoms, 3), dtype=numpy.float32)
+    for i in range(n_atom):
+        a = mol.GetAtomWithIdx(i)
+        if a.GetHybridization() is None:
+            continue
+        hybridization_type = str(a.GetHybridization())
+        if hybridization_type == 'SP1':
+            hybridization_vec[i, 0] = 1.0
+        elif hybridization_type == 'SP2':
+            hybridization_vec[i, 1] = 1.0
+        elif hybridization_type == 'SP3':
+            hybridization_vec[i, 2] = 1.0
+    return hybridization_vec
+
+
+def construct_partial_charge_vec(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    AllChem.ComputeGasteigerCharges(mol)
```

Comment:
If my understanding is correct, this function has a side effect on `mol`. If so, please note that.

Reply:
Sorry, which method?

Reply:
I intended `construct_partial_charge_vec`.

Reply:
`ComputeGasteigerCharges` changes `mol` ?

Reply:
Yes, [the document](http://www.rdkit.org/Python_Docs/rdkit.Chem.rdPartialCharges-module.html#ComputeGasteigerCharges) says it adds `_GasteigerCharge` to each atom in the `mol` instance.

```
from rdkit import Chem
from rdkit.Chem import AllChem


def f(mol):
    AllChem.ComputeGasteigerCharges(mol)


mol = Chem.MolFromSmiles('COc1cccc2cc(C(=O)NCCCCN3CCN(c4cccc5nccnc54)CC3)oc21')

assert not mol.GetAtomWithIdx(0).HasProp('_GasteigerCharge')
f(mol)
assert mol.GetAtomWithIdx(0).HasProp('_GasteigerCharge')
assert isinstance(mol.GetAtomWithIdx(0).GetProp('_GasteigerCharge'), str)
```